### PR TITLE
[DONE] Improved Accessbility in Song Data Grid

### LINKED ISSUE #809: Better accessbility while playing a song from an album
## Problem Description

The current way of playing a single song might be a bit difficult for some users to figure out since there is no distinct play button visible.

Screenshot from 2021-03-10 13-14-51

## Possible Solution

We can show a play button on hover next to the song name(kind of like what Spotify does in its web player. I have attached a screenshot for reference.

Screenshot from 2021-03-10 13-10-11

## References

https://open.spotify.com/

### PR DESCRIPTION
**Pull request by aniketbiswas21**
Screenshot from 2021-03-12 14-00-12

Fixes #809 

@deluan Currently, the button shows accordingly if a song is on the playing stage or paused stage. But, I have not yet implemented the pause feature(It restarts the song if you try to click on the pause state). So, I believe in order to implement the pause/resume feature in this, I would have to make the ```audioInstance``` present in ```Player.js``` to be accessible globally. This is what I thought of. I wanted to know if this would be the right way to do this or am I doing something wrong? I would be glad if you could just point me in the right direction.

Thanks!

### CODE REVIEW

**Comment by deluan** [comment]
Thanks for taking a stab at this! 

We should _not_ make the `audioInstance` globally available. The player is controlled in a declarative way, using Redux. To control the player you need to dispatch actions using Redux, take a look at the [playQueueReducer](https://github.com/navidrome/navidrome/blob/2397a7e464498db0bec7442bd4bd19afea1503d4/ui/src/reducers/playQueue.js#L49). 

We may need to introduce a `playing` boolean in the playQueue state, and then when we want to play/pause, we can dispatch a `togglePlay` action, which would flip the value of this flag. In the `Player.js`, we would act on this value when setting the [player options](https://github.com/navidrome/navidrome/blob/bc609be3c979ed3c07661feb52e28cd41247a730/ui/src/audioplayer/Player.js#L168).

Regarding the visuals, I'd like for this play/pause button to replace the number of the track when hovering over the track, similar to Spotify. Currently there is a small animation showing which track is being played. It appears in front of the track name, but I think we should also put this animation over the track number.

**Comment by aniketbiswas21** [comment]
@deluan I added the UI changes you requested. As for the play/pause toggle, I did make a ```togglePlayAction``` function in ```playQueue``` state. But, I could not figure out a way to access ```playing``` boolean in ```Player.js```. I tried passing it in the options like the other states but it was returning ```undefined```. Can you help me with this? Also, for listening on to the former  field,  would triggering a ```useEffect```  hook with ```playing``` as the ```dependency```  be a good option?

**Comment by aniketbiswas21** [comment]
@deluan I tried accessing the queue state like this: 
```const queue = useSelector((state)=>(state.queue));```
But, the property I added in the reducer was not showing up.

**Comment by deluan** [comment]
I'll take a look tomorrow. Maybe this is not as simple as it seems. Controlling the `react-jinke-music-player` sometimes is a pain, even though it blends well in Navidrome UI and is feature-packed.

**Comment by aniketbiswas21** [comment]
@deluan Ok :+1:

**Comment by deluan** [comment]
@aniketbiswas21 I see that you are using javascript to handle the hover events. This [used to be the case](https://github.com/navidrome/navidrome/commit/a42fb024be2da564e25b0bea09df9bb778d3a300) in all places I need to change elements on hover (ex: Context Menus), but it was slowing down the UI for really large song lists, ex: showing a playlist with >600 songs. I decided to move to a CSS-only solution for handling the hover.

Unfortunately, I won't be able to merge this if we don't at least try to use a CSS-only solution. I've been very mindful about adding these small things that may have a big impact in performance when displaying large amounts of data. At least while we don't have infinite scroll.

Regarding the Redux issue: I don't see you calling the `togglePlayAction` you created when you actually play/pause a song. That's why it is always empty/non-existent. Anyway, you may not need to use it, as I just remember we have a `current.paused` value in our state to track whether a song is being played or not (it is used for the animated icon showing the current song being played. So you may get rid of your action/reducer and just use it?

**Comment by aniketbiswas21** [comment]
@deluan I can implement a CSS-only solution if we can create an extra column for that play button. Would that be okay? In order to replace the number, I had to keep a track of the current hovered ```id```, that's why had to use js :sweat_smile:

I will certainly look for a solution to get the current layout. 

As for the redux one, I did call the action, but was getting undefined as the state value so removed it :sweat_smile: . Although, I would look on to what you mentioned :+1: 

Thanks for helping me out :smile:

**Comment by deluan** [comment]
I'd like to avoid adding an extra column..... Can't you have two elements in the same cell (track number and play/pause button), when one is hidden the other is visible and vice-versa? That way you could add the button with the id on each cell, and only change its visibility on hover, similar to how the context menu works. I may be missing something here, but I think it is doable.

**Comment by aniketbiswas21** [comment]
@deluan Yeah, I think I can do that. I will give it a shot.

Thanks a lot :+1:

**Comment by aniketbiswas21** [comment]
@deluan Made all the requested changes :sweat_smile: . Migrated to a CSS only solution and also implemented the play/pause feature.

Please take a look :sweat_smile: 

https://user-images.githubusercontent.com/51146347/111140101-f1eff680-85a7-11eb-9368-db3c27969e41.mp4

**Comment by aniketbiswas21** [comment]
@deluan is the above implementation fine? Please let me know if I did something wrong 😅

Thank you.

**Comment by deluan** [comment]
Thanks for this! I'll take a look first thing in the morning

File: ui/src/common/SongTitleField.js [hunk 1]
```diff
@@ -25,14 +18,7 @@ const useStyles = makeStyles({
 })
 
 export const SongTitleField = ({ showTrackNumbers, ...props }) => {
-  const theme = useTheme()
```
Comment (deluan) [inline ui/src/common/SongTitleField.js:28]: Why did you remove the logic from here and spread it into other components (breaking the behaviour for other song lists)? Why not just extending this component?

File: ui/src/reducers/playQueue.js [hunk 2]
```diff
@@ -81,7 +81,13 @@ export const playQueueReducer = (previousState = initialState, payload) => {
       Object.keys(data).forEach((id) => {
         queue.push(mapToAudioLists(data[id]))
       })
-      return { ...previousState, queue, clear: false, playIndex: undefined }
+      return {
+        ...previousState,
+        queue,
+        clear: false,
+        playIndex: undefined,
+        playing: true,
```
Comment (deluan) [inline ui/src/reducers/playQueue.js:89]: Any particular reason why you introduced another value in the store, instead of using the `current.paused` (line 71 above)?

**Review by deluan — CHANGES_REQUESTED** [review]
Thanks for the PR. See my comments below. Besides these comments, I also found a few issues:

1) There is an error being thrown to the console, that can be fixed:
Screen Shot 2021-03-18 at 10 56 28 PM

2) Mobile view, Playlist view and  SongList view, all lost the "Playing/Paused" icon. Most probably because you removed the logic from `SongTitleField`, as I pointed out below.

3) There is a small alignment issue:
ezgif com-gif-maker (1)

4) Can you please use icons without the circle? A simple play arrow / pause icon is cleaner:
Screen Shot 2021-03-18 at 11 15 42 PM

File: ui/src/reducers/playQueue.js (hunk 2 above)
Comment (aniketbiswas21) [reply ui/src/reducers/playQueue.js:89]: I did this to avoid recursive calls being made. Since, the current ```actual play/pause``` depends on ```current.paused```. So, it is flipped everytime I change the status of the song. So, if I listen on to the former i.e. act on the changing of the ```current.paused```, when the togglePlayPause would run again and flip the value again and so forth, triggering a recursion. To avoid this, I introduced another value to listen on to the changes and act accordingly.

**Comment by aniketbiswas21** [comment]
@deluan  I will get working on the requested changes ASAP :+1:

**Comment by aniketbiswas21** [comment]
@deluan Made all the requested changes. Please check :sweat_smile:

**Comment by aniketbiswas21** [comment]
@deluan Sir, sorry to bother you again:sweat: . But, can you take a look and let me know if I did anything wrong?

**Comment by deluan** [comment]
Sorry, this is a big PR, and I'm kind of swamped right now. But not worries, will take a look as soon as I can, probably tomorrow.

**Comment by aniketbiswas21** [comment]
> Sorry, this is a big PR, and I'm kind of swamped right now. But not worries, will take a look as soon as I can, probably tomorrow.

Sure sir :+1:

**Comment by aniketbiswas21** [comment]
@deluan I rebased this with the master. Please take a look :)
